### PR TITLE
Transaction optimization

### PR DESCRIPTION
**Pull request by shargon**
- `Witnesses` length are checked during verification against `Signers`.
-  `GetScriptHashesForVerifying` was called twice during `VerifyStateDependent`.
- `if (hashes.Length != witnesses.Length)` was already called during `VerifyStateIndependent`.
- added `if (net_fee < 0) return VerifyResult.InsufficientFunds;` because if we doesn't have funds for pay the size fee, we don't need to execute the verification.

### CODE REVIEW

File: src/neo/Wallets/Wallet.cs [hunk 1]
```diff
@@ -241,7 +241,7 @@ public virtual WalletAccount Import(string nep2, string passphrase, int N = 1638
             return account;
         }
 
-        public Transaction MakeTransaction(TransferOutput[] outputs, UInt160 from = null, Signer[] cosigners = null)
+        public Transaction MakeTransaction(TransferOutput[] outputs, UInt160 from = null, Signer[] cosigners = null, DataCache snapshot = null)
```
Comment (erikzhang) [inline src/neo/Wallets/Wallet.cs:244]: Why do we need this?

Comment (shargon) [reply src/neo/Wallets/Wallet.cs:244]: It's for UT, in order to fake the balance without commit the snapshot

Comment (erikzhang) [reply src/neo/Wallets/Wallet.cs:244]: We shouldn't change the APIs for UT only.

Comment (shargon) [reply src/neo/Wallets/Wallet.cs]: Changed

**Review by erikzhang — APPROVED** [review]

**Pull request merged**